File: csharp/demo/demo24-volatile.cs
/*
 * THE VOLATILE KEYWORD
 */
using System;
using System.Threading;



class Demo24 : IRunnable
{
    public void run()
    {
        Global.isRunning = true;
        new Thread(doTask).Start();

        Thread.Sleep(6000);
        Global.isRunning = false;
    }


    private void doTask()
    {
        while (Global.isRunning)
        {
            Console.WriteLine("Running...");
            Thread.Sleep(2000);
        }
    }


    class Global
    {
        public static volatile bool isRunning;
    }
}

File: cpp/cpp-pthread/demo24-volatile.cpp
/*
THE VOLATILE KEYWORD
*/


#include <iostream>
#include <unistd.h>
#include <pthread.h>
using namespace std;



volatile bool isRunning;



void* doTask(void*) {
    while (isRunning) {
        cout << "Running..." << endl;
        sleep(2);
    }

    pthread_exit(nullptr);
    return nullptr;
}



int main() {
    pthread_t tid;
    int ret = 0;

    isRunning = true;
    ret = pthread_create(&tid, nullptr, &doTask, nullptr);

    sleep(6);
    isRunning = false;

    ret = pthread_join(tid, nullptr);
    return 0;
}
